### PR TITLE
reimplement ngettext, fixes for gettext

### PR DESCRIPTION
**Pull request by bholloway**
I am proposing that we:
* merge this implementation
* Release some patch (?) that is < `1.0.0`
* Confirm it is good for our use-case
* Release a `1.0.0`

We could do an `npm link` if we want to be certain but I think this is a good increment either way.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,32 +1,46 @@
...
+  toToken = defaultToToken,
+  NODE_ENV = (process.env || {}).NODE_ENV
 } = {}) => (strings, ...substitutions) => {
+  // get a msgid template with sensible token names
   const tokens = substitutions.map(toToken);
   const {names, values, msgid} = getTemplate(strings, tokens);
-  assertTokens(tokens, `Error in gettext(${msgid})`);
+
+  // validate unless production
+  if (NODE_ENV !== 'production') {
```
Comment (nkbt) [inline src/index.js]: You must do process.env.node-env explicitly. It is pure text replacement by build tools. It will not infer value with this code.

**Review by nkbt — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 2]
```diff
@@ -43,58 +57,52 @@ export const gettextFactory = ({
...
  * @param {function} [toToken] Optional custom token inference function, yeilds {name, key, value}
- * @param {string} [delimiter] Optional custom delimiter character
- * @returns {function():function} Factory for a template string interpolator
+ * @param {function} [splitPlural] Optional split string to plural forms, yeilds Array
+ * @param {string} [NODE_ENV] Reserved for testing
+ * @returns {function(quantity:int):function} Factory for a template string interpolator
  */
 export const ngettextFactory = ({
-  ngettext = x => x,
+  ngettext = (s, p, q) => ((typeof q !== 'number') || isNaN(q) || (q === 1) ? s : p),
```
Comment (nkbt) [inline src/index.js]: That looks like default calues being quite a bit overused here. Why not just extract this as separate function? You probably would like to test it too then

File: src/plurals.js [hunk 3]
```diff
@@ -26,3 +26,14 @@ export const calculateDelimiters = delimiter => ({strings, msgstr, names, values
 
   return {lengthUntranslated, lengthTranslated, groups};
 };
+
+
+export const defaultSplitPlural = msgid => msgid.split('|');
+defaultSplitPlural.expect = 2;
```
Comment (nkbt) [inline src/plurals.js]: Why is this needed? And why is it static property of function?

**Review by nkbt — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 1 above)
Comment (bholloway) [reply src/index.js]: I think you need to move this comment to line 32.

If you use the `DefinePlugin` properly it will replace `node.env`. On reflection I will not take the risk, given there is now an `EnvironmentPlugin` for Webpack2 and I do not know how that behaves.

So the default value on line 32 will become `process.env.NODE_ENV` as suggested.

File: src/index.js (hunk 2 above)
Comment (bholloway) [reply src/index.js]: can do

File: src/plurals.js (hunk 3 above)
Comment (bholloway) [reply src/plurals.js]: Validate the correct number of delimiters. Don't want `splitPlural` to return Error since that wont work for production. Could move it to another `option` if that is less weird.

**Comment by nkbt** [comment]
Sure you can configure webpack for everything, but there are certain common
practices already in place. Check the source code of react, for example.

File: src/plurals.js (hunk 3 above)
Comment (nkbt) [reply src/plurals.js]: This looks like constant to me tbh

**Comment by bholloway** [comment]
Thats what I thought earlier today. Actually I could only find [one reference](https://github.com/facebook/react/blob/b1b4a2fb252f26fe10d29ba60d85ff89a85ff3ec/src/shared/utils/flattenChildren.js#L24).

There are also lots of tests that mutate `process.env` which I really don't want to do.

File: src/index.js (hunk 1 above)
Comment (bholloway) [reply src/index.js]: This has been fixed in the function arguments (i.e. line 32)

File: src/token.js [hunk 4]
```diff
@@ -1,33 +1,50 @@
...
-        (following.key === token.key) && (following.value !== token.value)
-      ))
-    .map(token => token.key);
-
-  if (errors.length) {
-    throw new Error(`${message}: substitution must be unique or have unique key: ${errors}`);
+  const collisions = calculateCollisions(tokens);
+  if (collisions.length) {
+    throw new Error(
+      `${message}: substitution with the same bane must have the same value: ${collisions}`);
```
Comment (nkbt) [inline src/token.js]: I don't understand why same name substitutions must have the same value. Why do you care about it?

```js
gettext`bla=${{bla: 'bla'}} foo=${{bla: 'foo'}}`

// would produce  --> `bla=__bla__ foo=__bla__` template
// but then replace into `bla=bla foo=foo` string
```

Though I may miss some implementation details and this is not actually possible...

**Review by nkbt — APPROVED** [review]

**Comment by nkbt** [comment]
Can't see why not merge this, looks good anyway


PS: where is your CCI build?

File: src/token.js (hunk 4 above)
Comment (bholloway) [reply src/token.js]: This is fine for `gettext` but not for `ngettext`.

There are some different solutions but they are less elegant/consistent. It is easier to be opinionated about it and avoid the problem.

Comment (bholloway) [reply src/token.js]: actually... your question highlights a problem that escaped the unit tests

Comment (nkbt) [reply src/token.js]: lol, happy to help =)

**Comment by bholloway** [comment]
(Sorry about the drip-feed of commits)
I have done `npm link` and integrated in the main project that I won't mention here.
Works for the `Paginator` use case.
Feel free to wait to see that that PR if you would like. It will need us to publish `0.2.0`.

**Comment by bholloway** [comment]
@nkbt Did you want to revisit this or just merge? There are some fixes that were necessary.
